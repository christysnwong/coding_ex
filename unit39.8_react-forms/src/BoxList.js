import React, {useState} from "react";
import Box from "./Box"
import NewBoxForm from "./NewBoxForm"


const BoxList = () => {
    const INITIAL_STATE = []

    const [boxes, setBoxes] = useState(INITIAL_STATE);
    const [boxCount, setBoxCount] = useState(1);

    const addBox = (newBox) => {

        setBoxes(boxes => [...boxes, {...newBox, id: boxCount}]);
        setBoxCount((boxCount) => boxCount + 1);
        
    }

    const removeBox = (id) => {
        setBoxes(boxes => boxes.filter(box => box.id !== id));
    }


    return (
      <div>
        <h3>Box List</h3>
        <NewBoxForm addBox={addBox} />
        <div>
            {boxes.map( ({id, width, height, bgColor}) => {
                return (
                  <Box
                    id={id}
                    key={id}
                    width={width}
                    height={height}
                    bgColor={bgColor}
                    removeBox={removeBox}
                  />
                );
            })}
        </div>
      </div>
    );
}

export default BoxList;