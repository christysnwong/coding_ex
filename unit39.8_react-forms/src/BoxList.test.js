import { render, fireEvent } from "@testing-library/react";
import BoxList from "./BoxList";

it("renders without crashing", function () {
  render(<BoxList />);
});

it("matches snapshot", function () {
  const { asFragment } = render(<BoxList />);
  expect(asFragment()).toMatchSnapshot();
});


it("should add new boxes", function () {
  const { queryByText, queryAllByText, getByLabelText } = render(<BoxList />);

  const input = getByLabelText("Color");
  const btn = queryByText("Add Box");

  expect(queryByText("1")).not.toBeInTheDocument();

  fireEvent.change(input, { target: { value: "red" } });
  fireEvent.click(btn);
  fireEvent.click(btn);

  const divBox1 = queryByText("1");
  const divBox2 = queryByText("2");

  expect(queryByText("1")).toBeInTheDocument();
  expect(queryByText("2")).toBeInTheDocument();

  expect(divBox1).toHaveStyle(`
    width: 150px;
    height: 150px;
    background-color: red;
  `);

  expect(divBox2).toHaveStyle(`
    width: 150px;
    height: 150px;
    background-color: cyan;
  `);

  expect(queryAllByText("Remove")).toHaveLength(2);

});

it("should remove box", function() {
    const { queryByText } = render(<BoxList />);

    expect(queryByText("Remove")).not.toBeInTheDocument();

    const addBtn = queryByText("Add Box");
    fireEvent.click(addBtn);

    expect(queryByText("1")).toBeInTheDocument();
    expect(queryByText("Remove")).toBeInTheDocument();

    const removeBtn = queryByText("Remove");
    fireEvent.click(removeBtn);

    expect(queryByText("1")).not.toBeInTheDocument();
    expect(queryByText("Remove")).not.toBeInTheDocument();


})