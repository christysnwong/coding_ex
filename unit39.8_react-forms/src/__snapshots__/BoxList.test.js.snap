// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapshot 1`] = `
<DocumentFragment>
  <div>
    <h3>
      Box List
    </h3>
    <form>
      <label
        for="width"
      >
        Box Width 
      </label>
      <input
        id="width"
        max="200"
        min="100"
        name="width"
        placeholder="Box Width in pixels"
        type="range"
        value=""
      />
      <label
        for="height"
      >
        Box Height 
      </label>
      <input
        id="height"
        max="200"
        min="100"
        name="height"
        placeholder="Box Height in pixels"
        type="range"
        value=""
      />
      <label
        for="bgcolor"
      >
        Color 
      </label>
      <input
        id="bgcolor"
        name="bgColor"
        placeholder="Colour of the Box"
        type="text"
        value=""
      />
      <button>
        Add Box
      </button>
    </form>
    <div />
  </div>
</DocumentFragment>
`;
